odl/space/base_tensors.py

```diff
@@ -0,0 +1,651 @@
+        >>> z in spc
+        False
+        >>> x in other_dtype_spc
+        False
+
+        On the other hand, spaces are not unique:
+
+        >>> spc2 = odl.tensor_space((2, 3), dtype='uint64')  # same as spc
+        >>> x2 = spc2.element([[5, 4, 3],
+        ...                    [2, 1, 0]])
+        >>> x2 in spc
+        True
+        >>> x in spc2
+        True
+        """
+        return getattr(other, 'space', None) == self
+
+    def __eq__(self, other):
+        """Return ``self == other``.
+
+        Returns
+        -------
+        equals : bool
+            True if ``other`` is an instance of ``type(self)`` with the
+            same `shape`, `dtype` and `order`, otherwise `False`.
+
+        Examples
+        --------
+        Sizes, data types and other essential properties characterize
+        spaces and decide about equality:
+
+        >>> spc = odl.tensor_space((2, 3), dtype='uint64')
+        >>> spc == spc
+        True
+        >>> spc2 = odl.tensor_space((2, 3), dtype='uint64')
+        >>> spc2 == spc
+        True
+        >>> smaller_spc = odl.tensor_space((2, 2), dtype='uint64')
+        >>> spc == smaller_spc
+        False
+        >>> other_dtype_spc = odl.tensor_space((2, 3), dtype='uint32')
+        >>> spc == other_dtype_spc
+        False
+        """
+        if other is self:
+            return True
+
+        return ((isinstance(self, type(other)) or
+                 isinstance(other, type(self))) and
+                self.shape == other.shape and
+                self.dtype == other.dtype and
+                self.order == other.order)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        posargs = [self.shape, dtype_str(self.dtype)]
+        optargs = [('order', self.order, self.default_order())]
+        return "{}({})".format(self.__class__.__name__,
+                               repr_signature(posargs, optargs))
+
+    @staticmethod
+    def default_order():
+        """Return the default axis ordering of this implementation."""
+        raise NotImplementedError('abstract method')
+
+    @staticmethod
+    def available_dtypes():
+        """Return the list of data types available in this implementation."""
+        raise NotImplementedError('abstract method')
+
+    @property
+    def element_type(self):
+        """Type of elements in this set: `BaseGeneralizedTensor`."""
+        return BaseGeneralizedTensor
+
+
+class BaseGeneralizedTensor(object):
+
+    """Abstract class for representation of `BaseTensorSet` elements.
+
+    Defines abstract and concrete attributes independent of data
+    representation.
+    """
+
+    def __init__(self, space, *args, **kwargs):
+        """Initialize a new instance."""
+        self.__space = space
+        self.__ufunc = BaseGeneralizedTensorUfunc(self)
+
+    def copy(self):
+        """Create an identical (deep) copy of this tensor."""
+        raise NotImplementedError('abstract method')
+
+    def asarray(self, out=None):
```

Comment:
Actually, I would keep the method because of its `out` parameter. GPU arrays have a special method to support that.